binary_search_tree/tree.py

```diff
@@ -1,59 +1,124 @@
         self.right = None
-        
 
 
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
-    def add(self, key, value = None):
-        pass
+    # Time Complexity: O(log n) balanced / O(n) unbalanced
+    # Space Complexity: O(log n) balanced / O(n) unbalanced
+    def add(self, key, value=None):
+        new_node = TreeNode(key, value)
+
+        def add_helper(curr, new):
+            if curr is None:
+                return new
+            if new_node.key <= curr.key:
+                curr.left = add_helper(curr.left, new)
+            else:
+                curr.right = add_helper(curr.right, new)
+            return curr
+
+        self.root = add_helper(self.root, new_node)
+
+    # Time Complexity: O(log n) balanced / O(n) unbalanced
+    # Space Complexity: O(log n) balanced / O(n) unbalanced
 
-    # Time Complexity: 
-    # Space Complexity: 
     def find(self, key):
-        pass
+        def find_helper(curr, key):
+            if curr is None:
+                return None
+            if curr.key == key:
+                return curr.value
+            if key <= curr.key:
+                return find_helper(curr.left, key)
+            else:
+                return find_helper(curr.right, key)
 
-    # Time Complexity: 
-    # Space Complexity: 
+        return find_helper(self.root, key)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        def inorder_helper(curr, arr):
+            if curr is None:
+                return arr
+            inorder_helper(curr.left, arr)
+            arr.append({"key": curr.key, "value": curr.value})
+            inorder_helper(curr.right, arr)
+            return arr
+
+        return inorder_helper(self.root, [])
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def preorder(self):
-        pass
+        def preorder_helper(curr, arr):
+            if curr is None:
+                return arr
+            arr.append({"key": curr.key, "value": curr.value})
+            preorder_helper(curr.left, arr)
+            preorder_helper(curr.right, arr)
+            return arr
+
+        return preorder_helper(self.root, [])
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n
 
-    # Time Complexity: 
-    # Space Complexity:     
     def postorder(self):
-        pass
+        def postorder_helper(curr, arr):
+            if curr is None:
+                return arr
+            postorder_helper(curr.left, arr)
+            postorder_helper(curr.right, arr)
+            arr.append({"key": curr.key, "value": curr.value})
+            return arr
+
+        return postorder_helper(self.root, [])
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(log n) balanced / O(n) unbalanced
+    # Space Complexity: O(log n) balanced / O(n) unbalanced
     def height(self):
```

Comment:
👍  However the time complexity is *always* going to be O(n) because you have to visit each node.